File: Lab-1/Exc-2/ansi/ansi.go
package ansi

import (
	"fmt"
)

type makeEscSeq func(uint) string
type make2EscSeq func(uint, uint) string
type makeEscSeq8 func(uint8) string

type colorMap struct {
	Black   string
	Red     string
	Green   string
	Yellow  string
	Blue    string
	Magenta string
	Cyan    string
	White   string
}

type colorPalete struct {
	colorMap
	Bright colorMap
	Code   makeEscSeq8
}

type formatMap struct {
	Bold      string
	Faint     string
	Italic    string
	Underline string
	Inverse   string
}

type formatSpecifiers struct {
	Reset string
	formatMap
	Off formatMap
}

type clearingMap struct {
	ToEnd   string
	ToStart string
	All     string
}

type clearingSpecifiers struct {
	Screen     clearingMap
	Line       clearingMap
	Everything string
}

type cursorMemory struct {
	Save    string
	Restore string
}

type cursorMove struct {
	Up   makeEscSeq
	Down makeEscSeq
}

type cursorGo struct {
	XY make2EscSeq
	X  makeEscSeq
}

type cursorSpecifiers struct {
	Up     makeEscSeq
	Down   makeEscSeq
	Right  makeEscSeq
	Left   makeEscSeq
	Line   cursorMove
	Scroll cursorMove
	GoTo   cursorGo
	Memo   cursorMemory
}

var Fg = colorPalete{
	colorMap: colorMap{
		Black:   "\u001b[30m",
		Red:     "\u001b[31m",
		Green:   "\u001b[32m",
		Yellow:  "\u001b[33m",
		Blue:    "\u001b[34m",
		Magenta: "\u001b[35m",
		Cyan:    "\u001b[36m",
		White:   "\u001b[37m",
	},
	Bright: colorMap{
		Black:   "\u001b[90m",
		Red:     "\u001b[91m",
		Green:   "\u001b[92m",
		Yellow:  "\u001b[93m",
		Blue:    "\u001b[94m",
		Magenta: "\u001b[95m",
		Cyan:    "\u001b[96m",
		White:   "\u001b[97m",
	},
	Code: func(c uint8) string { return fmt.Sprintf("\u001b[38;5;%vm", c) },
}

var Bg = colorPalete{
	colorMap: colorMap{
		Black:   "\u001b[40m",
		Red:     "\u001b[41m",
		Green:   "\u001b[42m",
		Yellow:  "\u001b[43m",
		Blue:    "\u001b[44m",
		Magenta: "\u001b[45m",
		Cyan:    "\u001b[46m",
		White:   "\u001b[47m",
	},
	Bright: colorMap{
		Black:   "\u001b[100m",
		Red:     "\u001b[101m",
		Green:   "\u001b[102m",
		Yellow:  "\u001b[103m",
		Blue:    "\u001b[104m",
		Magenta: "\u001b[105m",
		Cyan:    "\u001b[106m",
		White:   "\u001b[107m",
	},
	Code: func(c uint8) string { return fmt.Sprintf("\u001b[48;5;%vm", c) },
}

var Fmt = formatSpecifiers{
	Reset: "\u001b[0m",
	formatMap: formatMap{
		Bold:      "\u001b[1m",
		Faint:     "\u001b[2m",
		Italic:    "\u001b[3m",
		Underline: "\u001b[4m",
		Inverse:   "\u001b[7m",
	},
	Off: formatMap{
		Bold:      "\u001b[22m",
		Faint:     "\u001b[22m",
		Italic:    "\u001b[23m",
		Underline: "\u001b[24m",
		Inverse:   "\u001b[27m",
	},
}

var Clr = clearingSpecifiers{
	Screen: clearingMap{
		ToEnd:   "\u001b[0J",
		ToStart: "\u001b[1J",
		All:     "\u001b[2J",
	},
	Line: clearingMap{
		ToEnd:   "\u001b[0K",
		ToStart: "\u001b[1K",
		All:     "\u001b[2K",
	},
	Everything: "\u001b[3J",
}

var Cur = cursorSpecifiers{
	Up:    func(n uint) string { return fmt.Sprintf("\u001b[%vA", n) },
	Down:  func(n uint) string { return fmt.Sprintf("\u001b[%vB", n) },
	Right: func(n uint) string { return fmt.Sprintf("\u001b[%vC", n) },
	Left:  func(n uint) string { return fmt.Sprintf("\u001b[%vD", n) },
	Line: cursorMove{
		Up:   func(n uint) string { return fmt.Sprintf("\u001b[%vF", n) },
		Down: func(n uint) string { return fmt.Sprintf("\u001b[%vE", n) },
	},
	Scroll: cursorMove{
		Up:   func(n uint) string { return fmt.Sprintf("\u001b[%vS", n) },
		Down: func(n uint) string { return fmt.Sprintf("\u001b[%vT", n) },
	},
	GoTo: cursorGo{
		XY: func(y, x uint) string { return fmt.Sprintf("\u001b[%v;%vH", y, x) },
		X:  func(x uint) string { return fmt.Sprintf("\u001b[%vG", x) },
	},
	Memo: cursorMemory{
		Save:    "\u001b[s",
		Restore: "\u001b[u",
	},
}

File: Lab-1/src/main/kotlin/com/marcinjeznach/ansi/ansi.kt
@file:Suppress("ClassName", "ConstPropertyName")

package com.marcinjeznach.ansi

class Modifier(private val escapeSeq: String) {
	override fun toString() = escapeSeq

	/**
	 * Wraps message into this format.
	 *
	 * Clears all formatting at the end of returned string.
	 */
	operator fun get(msg: Any?) = "$escapeSeq$msg\u001B[0m"

	/** Produces modifier that adds [next] effect. */
	operator fun plus(next: Modifier) = Modifier(escapeSeq + next.escapeSeq)
}

/** Changes text color. */
object fg {
	val black = Modifier("\u001b[30m")
	val red = Modifier("\u001b[31m")
	val green = Modifier("\u001b[32m")
	val yellow = Modifier("\u001b[33m")
	val blue = Modifier("\u001b[34m")
	val magenta = Modifier("\u001b[35m")
	val cyan = Modifier("\u001b[36m")
	val white = Modifier("\u001b[37m")

	object br {
		val black = Modifier("\u001b[90m")
		val red = Modifier("\u001b[91m")
		val green = Modifier("\u001b[92m")
		val yellow = Modifier("\u001b[93m")
		val blue = Modifier("\u001b[94m")
		val magenta = Modifier("\u001b[95m")
		val cyan = Modifier("\u001b[96m")
		val white = Modifier("\u001b[97m")
	}

	/** Creates text color changer from 8-bit color code. */
	fun code(c: Int) = Modifier("\u001b[38;5;${c}m")
}

/** Changes background color. */
object bg {
	val black = Modifier("\u001b[40m")
	val red = Modifier("\u001b[41m")
	val green = Modifier("\u001b[42m")
	val yellow = Modifier("\u001b[43m")
	val blue = Modifier("\u001b[44m")
	val magenta = Modifier("\u001b[45m")
	val cyan = Modifier("\u001b[46m")
	val white = Modifier("\u001b[47m")

	object br {
		val black = Modifier("\u001b[100m")
		val red = Modifier("\u001b[101m")
		val green = Modifier("\u001b[102m")
		val yellow = Modifier("\u001b[103m")
		val blue = Modifier("\u001b[104m")
		val magenta = Modifier("\u001b[105m")
		val cyan = Modifier("\u001b[106m")
		val white = Modifier("\u001b[107m")
	}

	/** Creates background color changer from 8-bit color code. */
	fun code(c: Int) = Modifier("\u001b[48;5;${c}m")
}

/** Formats text. */
object fmt {
	/** Removes all formatting, including text and background color. */
	const val reset = "\u001b[0m"
	val bold = Modifier("\u001b[1m")
	val faint = Modifier("\u001b[2m")
	val italic = Modifier("\u001b[3m")
	val underline = Modifier("\u001b[4m")

	/** Inverts text and background color. */
	val inverse = Modifier("\u001b[7m")

	/** Removes specific text formatting. */
	object off {
		val bold = Modifier("\u001b[22m")
		val faint = Modifier("\u001b[22m")
		val italic = Modifier("\u001b[23m")
		val underline = Modifier("\u001b[24m")
		val inverse = Modifier("\u001b[27m")
	}
}

/** Clears text. */
object clr {
	/** Clears portion of screen. */
	object scr {
		const val toEnd = "\u001b[0J"
		const val toStart = "\u001b[1J"
		const val all = "\u001b[2J"
	}

	/** Clears portion of current line. */
	object ln {
		const val toEnd = "\u001b[0K"
		const val toStart = "\u001b[1K"
		const val all = "\u001b[2K"
	}

	/** Clears screen and scroll history. */
	const val everything = "\u001b[3J"
}

/** Changes cursor position */
object cur {
	fun up(n: Int) = "\u001b[${n}A"
	fun down(n: Int) = "\u001b[${n}B"
	fun right(n: Int) = "\u001b[${n}C"
	fun left(n: Int) = "\u001b[${n}D"

	/** Changes cursor position, returning to beginning of the line. */
	object ln {
		fun up(n: Int) = "\u001b[${n}F"
		fun down(n: Int) = "\u001b[${n}E"
	}

	/**
	 * Scrolls screen buffer in specified direction, moving cursor in reverse direction.
	 *
	 * After change, cursor stays in same place - only printed content shifts,
	 * replacing old one, instead of moving it like [println()][println] does.
 	 */

	object scr {
		fun up(n: Int) = "\u001b[${n}S"
		fun down(n: Int) = "\u001b[${n}T"
	}

	/**
	 * Moves cursor to specified absolute position.
	 */
	object goTo {
		fun yx(y: Int, x: Int) = "\u001b[${y};${x}H"
		fun x(x: Int) = "\u001b[${x}G"
	}

	/**
	 * Saves / restores cursor position.
	 *
	 * For terminals with x-term emulation, use [xmemo].
	 */
	object memo {
		const val save = "\u001b[s"
		const val restore = "\u001b[u"
	}

	/**
	 * Saves / restores cursor position.
	 *
	 * For terminals with non-x-term emulation, use [memo].
	 */
	object xmemo {
		const val save = "\u001b7"
		const val restore = "\u001b8"
	}
}
